lib/db/mysql/index.js

```diff
@@ -606,6 +610,23 @@ MysqlStore.prototype = {
       .then(this._write.bind(this, QUERY_CODE_DELETE_USER, [id]));
   },
 
+  /**
+   * Removes user's tokens and refreshTokens for canGrant and publicClient clients
+   *
+   * @param userId
+   * @returns {Promise}
+   */
+  removePublicAndCanGrantTokens: function removePublicAndCanGrantTokens(userId) {
+    const uid = buf(userId);
+
+    return this._read(QUERY_PUBLIC_CLIENTS_LIST).then((_clients) => {
+      const clientIds = _clients.map((client) => client.id);
+
+      return this._write(QUERY_DELETE_ACCESS_TOKEN_FOR_PUBLIC_CLIENTS, [uid, clientIds])
```

Comment:
Looking up the client ids in a separate query is fine, but you could skip a round-trip by using MySQL's neat ability to delete through a join:

  https://stackoverflow.com/questions/652770/delete-with-join-in-mysql#4192849

It'd look something like (untested):

```
DELETE tokens
FROM tokens INNER JOIN clients ON tokens.clientId = clients.id
WHERE tokens.userId = ? AND (clients.publicClient = 1 OR clients.canGrant = 1)
```

Reply:
> MySQL's neat ability

Let's stay away from that 🤕  IMHO